cardano-cli/src/Cardano/CLI/Mary/Parser.hs

```diff
@@ -0,0 +1,216 @@
+  | NegativeLovelaceValueError
+    -- ^ Parsed 'Value' consists of a negative amount of lovelace.
+      !Quantity
+      -- ^ Lovelace quantity which caused the error.
+  deriving Show
+
+-- | Render an error message for a 'ValueParseError'.
+renderValueParseError :: ValueParseError -> Text
+renderValueParseError err =
+  case err of
+    ValueParseError pErr -> renderParsecParseError pErr
+    NegativeLovelaceValueError x ->
+      "Lovelace value must be positive, but it was "
+        <> Text.pack (show x)
+        <> "."
+
+-- | Render an error message for a Parsec 'ParseError'.
+--
+-- TODO: Improve error message
+renderParsecParseError :: ParseError -> Text
+renderParsecParseError =
+  intercalate ", "
+    . map (Text.pack . messageString)
+    . errorMessages
+
+------------------------------------------------------------------------------
+-- Expression parser
+------------------------------------------------------------------------------
+
+-- | Intermediate representation of a parsed multi-asset value.
+data ValueExpr
+  = ValueExprAdd !ValueExpr !ValueExpr
+  | ValueExprLovelace !Quantity
+  | ValueExprMultiAsset !PolicyId !AssetName !Quantity
+  deriving (Eq, Ord, Show)
+
+expression :: Parser ValueExpr
+expression = buildExpressionParser operatorTable term <?> "expression"
+
+tokenParser :: GenTokenParser String u Identity
+tokenParser = makeTokenParser haskellDef -- TODO: What language def to use?
+
+term :: Parser ValueExpr
+term =
+  parens tokenParser expression
+    <|> value
+    <?> "multi-asset value expression"
+
+operatorTable :: [[Operator String u Identity ValueExpr]]
+operatorTable =
+  [ [prefix "-" negateValueExpr]
+  , [binary "+" ValueExprAdd AssocLeft]
+  ]
+
+binary :: String -> (a -> a -> a) -> Assoc -> Operator String u Identity a
+binary name fun = Infix (reservedOp tokenParser name $> fun)
+
+prefix :: String -> (a -> a) -> Operator String u Identity a
+prefix name fun = Prefix (reservedOp tokenParser name $> fun)
+
+-- | Negate a 'ValueExpr'.
+negateValueExpr :: ValueExpr -> ValueExpr
+-- TODO @intricate: Consider how to properly handle this case.
+negateValueExpr (ValueExprAdd _x _y) = error "Negation of a value addition expression is not supported."
+negateValueExpr (ValueExprLovelace x) = ValueExprLovelace (negate x)
+negateValueExpr (ValueExprMultiAsset polId aName x) =
+  ValueExprMultiAsset polId aName (negate x)
+
+-- | Parse either a 'ValueExprLovelace' or 'ValueExprMultiAsset'.
+value :: Parser ValueExpr
+value = do
+  q <- try quantity <?> "quantity (word64)"
+  _ <- some space
+  aId <- assetId
+  _ <- spaces
+  pure $ case aId of
+    AdaAssetId -> ValueExprLovelace q
+    AssetId polId aName -> ValueExprMultiAsset polId aName q
+
+------------------------------------------------------------------------------
+-- Primitive parsers
+------------------------------------------------------------------------------
+
+-- | Period (\".\") parser.
+period :: Parser String
+period = string "."
```

Comment:
If we make this have type `Parser ()` then when we use it we don't need the `_ <- ` syntax.

Reply:
Oh, that's true. Thanks.